sim/simulator.py

```diff
@@ -21,6 +21,79 @@
     from .maze import ExtraCellInfo
 
 
+class Counter():
+    """A counter for various robot actions."""
+
+    def __init__(self):
+        self._current_step_counter = 0
+        self._total_step_counter = 0
+        self._best_step_counter = 0
+        self._current_weight_counter = 0
+        self._total_weight_counter = 0
+        self._best_weight_counter = 0
+
+    def count_action(self, action: Action):
+        """Update counters by the robot's action."""
+        self._current_step_counter += 1
+        self._total_step_counter += 1
+        match action:
+            case Action.READY:
+                weight = 0
+            case Action.RESET:
+                weight = 0
+                self.reset_current()
+            case Action.FORWARD:
+                weight = 1
+            case Action.BACKWARDS:
+                weight = 2
+            case Action.TURN_LEFT | Action.TURN_RIGHT:
+                weight = 3
+
+        self._current_weight_counter += weight
+        self._total_weight_counter += weight
+
+    def reset_current(self):
+        """Reset current counters and update best counters"""
+        if self._best_step_counter == 0 or self._best_step_counter > self._current_step_counter:
+            self._best_step_counter = self._current_step_counter
+
+        if self._best_weight_counter == 0 or self._best_weight_counter > self._current_weight_counter:
+            self._best_weight_counter = self._current_weight_counter
```

Comment:
This glitches with the `Reset` action: you'll count the way back as a "best" candidate...

I think it would be better to remove this logic from this function (and maybe also not call this function when reaching the goal)